### PR TITLE
[Fixes] VM - Update NIC IP Address handover

### LINKED ISSUE #1676: [Bug Report]: Cannot deploy Virtual Machine with Static IP Address
### Describe the bug

The `ipConfiguration.vmIPAddress` property within the Virtual Machine module is sent via a nested bicep resource to the Network Interfaces module using `privateIPAddress`, instead of `vmIPAddress`

I would assume we cannot update the Network Intefaces modules to align to the `privateIPAddress` property (as that would be a break public interface, so I guess the fix is to map the Virtual Machine nested bicep to use `vmIPAddress` throughout.

1st nested layer:
https://github.com/Azure/ResourceModules/blob/be7074418e312a8c6790b2879e2e669cb4e36aad/modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep#L68

2nd nested layer:
https://github.com/Azure/ResourceModules/blob/be7074418e312a8c6790b2879e2e669cb4e36aad/modules/Microsoft.Network/networkInterfaces/deploy.bicep#L107

### To reproduce

Example Bicep module call below

### Code snippet

```Bicep
module vm 'main/modules/Microsoft.Compute/virtualMachines/deploy.bicep' = {
  name: '${deployment().name}_vm'
  params: {
    name: 'name'
    imageReference: {
      publisher: 'zscaler'
      offer: 'zscaler-private-access'
      sku: 'zpa-con-azure'
      version: 'latest'
    }
    vmSize: 'Standard_D4s_v3'
    adminUsername: 'azure-username'
    adminPassword: 'supersecretpassword'
    osType: 'Linux'
    osDisk: {
      diskSizeGB: '16'
      managedDisk: {
        storageAccountType: 'StandardSSD_LRS'
      }
    }
    nicConfigurations: [
      {
        nicSuffix: '-nic-01'
        enableAcceleratedNetworking: true
        ipConfigurations: [
          {
            name: 'ipconfig01'
            subnetId: '/subscriptions/...../subnets/....'
            privateIPAllocationMethod: 'Static'
            vmIPAddress: '192.168.10.10'
          }
        ]
      }
    ]
  }
}
```

### Relevant log output

```Shell
ERROR: InvalidTemplateDeployment - The template deployment 'deploy1' is not valid according to the validation procedure. The tracking id is 'xxxxxxxxxxxxxxxxxxxxxxx'. See inner errors for details.
TemplateFailedPrevalidation - The template request has multiple validation failures involving one or more Microsoft.Network resources. Please check details for information about each resource.
PrivateIPAddressMissing - Private IP address is required when privateIPAllocationMethod is Static in IP configuration /subscriptions/xxxxxxxxxxxxxxxx/resourceGroups/xxxxxxxxxxxxxxx/providers/Microsoft.Network/networkInterfaces/name-nic-01/ipConfigurations/ipconfig01.
```

### PR DESCRIPTION
**Pull request by chris5287**
# Description

Fixes #1676 

The `ipConfiguration.vmIPAddress` property within the Virtual Machine module is sent via a nested bicep resource to the Network Interfaces module using `privateIPAddress`, instead of `vmIPAddress`

I would assume we cannot update the Network Intefaces modules to align to the `privateIPAddress` property (as that would be a break public interface, so I guess the fix is to map the Virtual Machine nested bicep to use `vmIPAddress` throughout.

1st nested layer:
https://github.com/Azure/ResourceModules/blob/be7074418e312a8c6790b2879e2e669cb4e36aad/modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep#L68

2nd nested layer:
https://github.com/Azure/ResourceModules/blob/be7074418e312a8c6790b2879e2e669cb4e36aad/modules/Microsoft.Network/networkInterfaces/deploy.bicep#L107

### To reproduce

Example Bicep module call below

```Bicep
module vm 'main/modules/Microsoft.Compute/virtualMachines/deploy.bicep' = {
  name: '${deployment().name}_vm'
  params: {
    name: 'name'
    imageReference: {
      publisher: 'zscaler'
      offer: 'zscaler-private-access'
      sku: 'zpa-con-azure'
      version: 'latest'
    }
    vmSize: 'Standard_D4s_v3'
    adminUsername: 'azure-username'
    adminPassword: 'supersecretpassword'
    osType: 'Linux'
    osDisk: {
      diskSizeGB: '16'
      managedDisk: {
        storageAccountType: 'StandardSSD_LRS'
      }
    }
    nicConfigurations: [
      {
        nicSuffix: '-nic-01'
        enableAcceleratedNetworking: true
        ipConfigurations: [
          {
            name: 'ipconfig01'
            subnetId: '/subscriptions/...../subnets/....'
            privateIPAllocationMethod: 'Static'
            vmIPAddress: '192.168.10.10'
          }
        ]
      }
    ]
  }
}
```

# Type of Change

Please delete options that are not relevant.

- [x] Bugfix (non-breaking change which fixes an issue)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep [hunk 1]
```diff
@@ -65,7 +65,7 @@ module networkInterface '../../../Microsoft.Network/networkInterfaces/deploy.bic
       name: !empty(ipConfiguration.name) ? ipConfiguration.name : null
       primary: index == 0
       privateIPAllocationMethod: contains(ipConfiguration, 'privateIPAllocationMethod') ? (!empty(ipConfiguration.privateIPAllocationMethod) ? ipConfiguration.privateIPAllocationMethod : null) : null
-      privateIPAddress: contains(ipConfiguration, 'vmIPAddress') ? (!empty(ipConfiguration.vmIPAddress) ? ipConfiguration.vmIPAddress : null) : null
+      vmIPAddress: contains(ipConfiguration, 'vmIPAddress') ? (!empty(ipConfiguration.vmIPAddress) ? ipConfiguration.vmIPAddress : null) : null
```
Comment (MrMCake) [inline modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:68]: Hey @chris5287, I would actually argue the opposite: We should update the NIC module as it's interface is incorrect.

Reviewing all properties I'd argue:
```Bicep
ipConfigurations: [for (ipConfiguration, index) in ipConfigurations: {
  name: !empty(ipConfiguration.name) ? ipConfiguration.name : null
  properties: {
    primary: index == 0 ? true : false
    privateIPAllocationMethod: contains(ipConfiguration, 'privateIPAllocationMethod') ? (!empty(ipConfiguration.privateIPAllocationMethod) ? ipConfiguration.privateIPAllocationMethod : null) : null
    privateIPAddress: contains(ipConfiguration, 'vmIPAddress') ? (!empty(ipConfiguration.vmIPAddress) ? ipConfiguration.vmIPAddress : null) : null
    publicIPAddress: contains(ipConfiguration, 'publicIPAddressResourceId') ? (ipConfiguration.publicIPAddressResourceId != null ? {
      id: ipConfiguration.publicIPAddressResourceId
    } : null) : null
    subnet: {
      id: ipConfiguration.subnetId
    }
    loadBalancerBackendAddressPools: contains(ipConfiguration, 'loadBalancerBackendAddressPools') ? ipConfiguration.loadBalancerBackendAddressPools : null
    applicationSecurityGroups: contains(ipConfiguration, 'applicationSecurityGroups') ? ipConfiguration.applicationSecurityGroups : null
    applicationGatewayBackendAddressPools: contains(ipConfiguration, 'applicationGatewayBackendAddressPools') ? ipConfiguration.applicationGatewayBackendAddressPools : null
    gatewayLoadBalancer: contains(ipConfiguration, 'gatewayLoadBalancer') ? ipConfiguration.gatewayLoadBalancer : null
    loadBalancerInboundNatRules: contains(ipConfiguration, 'loadBalancerInboundNatRules') ? ipConfiguration.loadBalancerInboundNatRules : null
    privateIPAddressVersion: contains(ipConfiguration, 'privateIPAddressVersion') ? ipConfiguration.privateIPAddressVersion : null
    virtualNetworkTaps: contains(ipConfiguration, 'virtualNetworkTaps') ? ipConfiguration.virtualNetworkTaps : null
  }
}]
```

should be changed to

```Bicep
ipConfigurations: [for (ipConfiguration, index) in ipConfigurations: {
  name: !empty(ipConfiguration.name) ? ipConfiguration.name : null
  properties: {
    primary: index == 0 ? true : false
    privateIPAllocationMethod: contains(ipConfiguration, 'privateIPAllocationMethod') ? (!empty(ipConfiguration.privateIPAllocationMethod) ? ipConfiguration.privateIPAllocationMethod : null) : null
    privateIPAddress: contains(ipConfiguration, 'privateIPAddress') ? (!empty(ipConfiguration.privateIPAddress) ? ipConfiguration.privateIPAddress : null) : null
    publicIPAddress: contains(ipConfiguration, 'publicIPAddressResourceId') ? (ipConfiguration.publicIPAddressResourceId != null ? {
      id: ipConfiguration.publicIPAddressResourceId
    } : null) : null
    subnet: {
      id: ipConfiguration.subnetResourceId
    }
    loadBalancerBackendAddressPools: contains(ipConfiguration, 'loadBalancerBackendAddressPools') ? ipConfiguration.loadBalancerBackendAddressPools : null
    applicationSecurityGroups: contains(ipConfiguration, 'applicationSecurityGroups') ? ipConfiguration.applicationSecurityGroups : null
    applicationGatewayBackendAddressPools: contains(ipConfiguration, 'applicationGatewayBackendAddressPools') ? ipConfiguration.applicationGatewayBackendAddressPools : null
    gatewayLoadBalancer: contains(ipConfiguration, 'gatewayLoadBalancer') ? ipConfiguration.gatewayLoadBalancer : null
    loadBalancerInboundNatRules: contains(ipConfiguration, 'loadBalancerInboundNatRules') ? ipConfiguration.loadBalancerInboundNatRules : null
    privateIPAddressVersion: contains(ipConfiguration, 'privateIPAddressVersion') ? ipConfiguration.privateIPAddressVersion : null
    virtualNetworkTaps: contains(ipConfiguration, 'virtualNetworkTaps') ? ipConfiguration.virtualNetworkTaps : null
  }
}]
``` 

effectively updating both the `privateIPAddress` & `subnet` property.
```

Comment (MrMCake) [reply modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:68]: Naturally that would mean we'd need to update also the NIC module's test file & the VM modules handover (as the `SubnetId` would now be `SubnetResourceId`). 

Would you agree?

Comment (chris5287) [reply modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:68]: Ah cool, I wasn't sure how you would feel about this as it's a breaking change for the current CARML interface.

Comment (MrMCake) [reply modules/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:68]: It sure is - but as we don't yet have a strict breaking change policy (pre V1.0) I'd argue now is the best time to make it 'right'.

File: modules/Microsoft.Network/networkInterfaces/deploy.bicep [hunk 2]
```diff
@@ -104,7 +104,7 @@ resource networkInterface 'Microsoft.Network/networkInterfaces@2021-05-01' = {
       properties: {
         primary: index == 0 ? true : false
         privateIPAllocationMethod: contains(ipConfiguration, 'privateIPAllocationMethod') ? (!empty(ipConfiguration.privateIPAllocationMethod) ? ipConfiguration.privateIPAllocationMethod : null) : null
-        privateIPAddress: contains(ipConfiguration, 'vmIPAddress') ? (!empty(ipConfiguration.vmIPAddress) ? ipConfiguration.vmIPAddress : null) : null
+        privateIPAddress: contains(ipConfiguration, 'privateIPAddress') ? (!empty(ipConfiguration.privateIPAddress) ? ipConfiguration.privateIPAddress : null) : null
```
Comment (MrMCake) [inline modules/Microsoft.Network/networkInterfaces/deploy.bicep:107]: Looks good. Just one thing: Please also update the test file in the `.test` subfolder - and the module's readme (i.e., deployment examples). This way the test pipelines will continue to work

Comment (MrMCake) [reply modules/Microsoft.Network/networkInterfaces/deploy.bicep:107]: You're right. The change would only apply to my other suggested change, that is, renaming `subnetId` to `subnetResourceId` in alignement with the `publicIPAddressResourceId` property. I'll take care of that in a subsequent PR

**Review by MrMCake — APPROVED** [review]

**Comment by MrMCake** [comment]
Thank you for your contribution @chris5287

**Pull request merged**